### ISSUE TITLE
log files are broken

### ISSUE THREAD

**Issue by albertz** [rwth-i6/sisyphus#229]
For example:
```
$ ls -la work/i6_experiments/users/zeyer/datasets/utils/extract_seq_list/ExtractSeqListJob.5LJd3Q5k1nEN/
total 189
drwxr-xr-x  6 zeyer assi  12 Dec  2 18:24 .
drwxr-xr-x 11 zeyer assi  11 Dec  2 15:08 ..
drwxr-xr-x  2 zeyer assi  12 Dec  3 05:00 engine
-rw-r--r--  1 zeyer assi 334 Dec  2 23:55 info
drwxr-xr-x  2 zeyer assi   3 Dec  2 18:24 input
-rw-r--r--  1 zeyer assi 900 Dec  2 23:55 job.save
lrwxrwxrwx  1 zeyer assi 112 Dec  2 18:24 log.run.1.0 -> engine/i6_experiments.users.zeyer.datasets.utils.extract_seq_list.ExtractSeqListJob.5LJd3Q5k1nEN.run.9589181.1.0
lrwxrwxrwx  1 zeyer assi 112 Dec  2 18:24 log.run.1.1 -> engine/i6_experiments.users.zeyer.datasets.utils.extract_seq_list.ExtractSeqListJob.5LJd3Q5k1nEN.run.9589181.1.1
drwxr-xr-x  2 zeyer assi   3 Dec  2 18:28 output
-rw-r--r--  1 zeyer assi 905 Dec  3 05:00 submit_log.run
-rw-r--r--  1 zeyer assi 600 Dec  2 18:50 usage.run.1
drwxr-xr-x  2 zeyer assi   2 Dec  2 18:24 work
```
However, in `engine`:
```
$ ls -la work/i6_experiments/users/zeyer/datasets/utils/extract_seq_list/ExtractSeqListJob.5LJd3Q5k1nEN/engine/
total 170
drwxr-xr-x 2 zeyer assi    12 Dec  3 05:00 .
drwxr-xr-x 6 zeyer assi    12 Dec  2 18:24 ..
-rw-r--r-- 1 zeyer assi  8778 Dec  2 18:33 i6_experiments.users.zeyer.datasets.utils.extract_seq_list.ExtractSeqListJob.5LJd3Q5k1nEN.run.9589181.1
-rw-r--r-- 1 zeyer assi  2173 Dec  2 18:38 i6_experiments.users.zeyer.datasets.utils.extract_seq_list.ExtractSeqListJob.5LJd3Q5k1nEN.run.9589181.1.batch
-rw-r--r-- 1 zeyer assi 10404 Dec  2 18:50 i6_experiments.users.zeyer.datasets.utils.extract_seq_list.ExtractSeqListJob.5LJd3Q5k1nEN.run.9589250.1
-rw-r--r-- 1 zeyer assi  2702 Dec  2 18:54 i6_experiments.users.zeyer.datasets.utils.extract_seq_list.ExtractSeqListJob.5LJd3Q5k1nEN.run.9589250.1.batch
```

The job is defined as follows:
```python
class ExtractSeqListJob(Job):
    """
    Takes any dataset dict, and extracts all seq tags from it.
    """

    def __init__(
        self,
        *,
        returnn_dataset: Dict[str, Any],  # to get all seq tags
        returnn_root: Optional[tk.Path] = None,
    ):
        self.returnn_dataset = returnn_dataset
        self.returnn_root = returnn_root

        self.out_seq_list = self.output_path("out_seq_list.txt")

        self.rqmt = {"cpu": 1, "mem": 4, "time": 1, "gpu": 0}

    def tasks(self):
        yield Task("run", rqmt=self.rqmt)

    def run(self):
        ...
```

Multiple things are messed up:

* Why are there `log.run.1.0` and `log.run.0.0`? This is not a batched/multi-node job. Maybe #212 messed this up?
* Those `log.run.*` files are not regular files but symlinks?
* Those `log.run.*` symlinks are broken?
* Those `log.run.*` symlinks are not updated to the latest Slurm job id?

**Comment by NeoLegends** [comment]
Can you show me the env variables from inside the job? The code after #212 will only try to link multiple logfiles if there is `SLURM_NTASKS > 1` or `SLURM_NPROCS > 1`.

**Comment by albertz** [comment]
So, I logged on some node where currently such a job is running. What I already find weird:
```
$ ps a --forest -u $(whoami) -o pid,comm,args
...
1752684 python3.11      /work/tools/users/zeyer/py-envs/py3.11-torch2.1/bin/python3.11 ./sis worker --engine long work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6 run
1752683 python3.11      /work/tools/users/zeyer/py-envs/py3.11-torch2.1/bin/python3.11 ./sis worker --engine long work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6 run
1752643 slurm_script    /bin/sh /var/spool/slurm/slurmd/job9595321/slurm_script
1752653  \_ srun         \_ srun -o work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6/engine/%x.%A.%a /work/tools/users/zeyer/py-envs/py3.11-torch2.1/bin/python3.11 ./sis worker --engine long work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6 run
1752655      \_ srun         \_ srun -o work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6/engine/%x.%A.%a /work/tools/users/zeyer/py-envs/py3.11-torch2.1/bin/python3.11 ./sis worker --engine long work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6 run
```
Why are there two procs (1752684 and 1752683)?

The env of 1752643 (slurm_script):
<details>

`PWD=/u/zeyer/setups/combined/2021-05-31^@DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/2574/bus^@HOME=/u/zeyer^@HOMEBREW_FORCE_BREWED_CURL=1^@INFOPATH=/u/zeyer/.linuxbrew/share/info:/u/zeyer/.linuxbrew/share/info:^@LANG=en_US.UTF-8^@LANGUAGE=en_US.UTF-8^@LC_ALL=en_US.UTF-8^@LC_IDENTIFICATION=en_GB.UTF-8^@LC_PAPER=en_GB.UTF-8^@LC_ADDRESS=en_GB.UTF-8^@LC_TELEPHONE=en_GB.UTF-8^@LC_CTYPE=en_US.UTF-8^@LC_TIME=en_GB.UTF-8^@TERM=tmux-256color^@XDG_SESSION_TYPE=tty^@SSH_TTY=/dev/pts/0^@TERM_PROGRAM=tmux^@TEMP=/var/tmp^@LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu:/usr/local/cudnn-11.X-v8.4/lib64:/usr/local/cuda-11.7/lib64:/usr/local/cuda-11.7/extras/CUPTI/lib64:/usr/local/cudnn-10.1-v7.6/lib64:/usr/local/cudnn-9.1-v7.1/lib64:/usr/local/cudnn-8.0-v7.0/lib64:/usr/local/cudnn-8.0-v6.0/lib64:/usr/local/cudnn-8.0-v5.1/lib64:/usr/local/cuda-9.1/lib64:/usr/local/cuda-9.1/extras/CUPTI/lib64:/usr/local/cuda-10.1/lib64:/usr/local/cuda-10.1/extras/CUPTI/lib64:/usr/local/cuda-10.0/lib64:/usr/local/cuda-10.0/extras/CUPTI/lib64:/usr/lib/atlas-base:/usr/local/cuda-7.5/lib64^@SHELL=/bin/zsh^@USER=zeyer^@PATH=/u/zeyer/.gem/ruby/1.9.1/bin:/u/zeyer/dotfiles/bin:/u/zeyer/dotfiles/system-tools/bin:/u/zeyer/tools/bin:/u/zeyer/.local/bin:/u/zeyer/.linuxbrew/bin:/u/zeyer/.gem/ruby/1.9.1/bin:/u/zeyer/dotfiles/bin:/u/zeyer/dotfiles/system-tools/bin:/u/zeyer/tools/bin:/u/zeyer/.local/bin:/u/zeyer/.linuxbrew/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts^@MOTD_SHOWN=pam^LOGNAME=zeyer^@MANPATH=/u/zeyer/.linuxbrew/share/man:/u/zeyer/.linuxbrew/share/man:^@ TERM_PROGRAM_VERSION=3.3a^@HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1^@MY_LOCAL_MACHINE=indigo^@SHLVL=2^@TMUX_PANE=%13^@SSH_CONNECTION=137.226.36.158 33678 137.226.36.63 22^@NCCL_DEBUG=INFO^@TMUX_PLUGIN_MANAGER_PATH=/u/zeyer/.tmux/plugins/^@LC_MEASUREMENT=en_GB.UTF-8^@LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=00:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.avif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:*~=00;90:*#=00;90:*.bak=00;90:*.old=00;90:*.orig=00;90:*.part=00;90:*.rej=00;90:*.swp=00;90:*.tmp=00;90:*.dpkg-dist=00;90:*.dpkg-old=00;90:*.ucf-dist=00;90:*.ucf-new=00;90:*.ucf-old=00;90:*.rpmnew=00;90:*.rpmorig=00;90:*.rpmsave=00;90:^@PAPERSIZE=a4^XDG_SESSION_ID=10509^@HOMEBREW_ARCH=core2^@LC_MONETARY=en_GB.UTF-8^@SSH_CLIENT=137.226.36.158 33678 22^@XDG_SESSION_CLASS=user^@XDG_RUNTIME_DIR=/run/user/2574^@TMUX=/tmp/tmux-2574/default,1180403,2^@TMPDIR=/var/tmp^@LC_NAME=en_GB.UTF-8^@LC_NUMERIC=en_GB.UTF-8^@OLDPWD=/u/zeyer/setups/combined/2021-05-31^@TMP=/var/tmp^@SLURM_JOB_NAME=i6_experiments.users.zeyer.datasets.utils.serialize.ReturnnDatasetToTextLinesJob.nVowScvUWQv6.run^@SLURM_PRIO_PROCESS=0^@SLURM_SUBMIT_DIR=/u/zeyer/setups/combined/2021-05-31^@SLURM_SUBMIT_HOST=indigo^@SLURM_CPUS_PER_TASK=1^@SLURM_ARRAY_TASK_COUNT=1^@SLURM_ARRAY_TASK_MIN=1^@SLURM_ARRAY_TASK_MAX=1^@SLURM_ARRAY_TASK_STEP=1^@SLURM_CLUSTER_NAME=cluster^@SLU RM_JOB_ID=9595321^@SLURM_JOB_NUM_NODES=1^@SLURM_ARRAY_JOB_ID=9595321^@SLURM_ARRAY_TASK_ID=1^@SLURM_JOB_NODELIST=cn-611^@SLURM_JOB_PARTITION=cpu_modern^@SLURM_NODE_ALIASES=(null)^@SLURM_JOB_CPUS_PER_NODE=2^@ENVIRONMENT=BATCH^@HOSTNAME=cn-611^@SLURM_JOBID=9595321^@SLURM_NNODES=1^@SLURM_NODELIST=cn-611^@SLURM_TASKS_PER_NODE=2^@SLURM_MEM_PER_NODE=20480^@SLURM_JOB_ACCOUNT=hlt^@SLURM_JOB_QOS=normal^@SLURM_TOPOLOGY_ADDR=cn-611^@SLURM_TOPOLOGY_ADDR_PATTERN=node^@SLURM_CONF=/var/spool/slurm/slurmd/conf-cache/slurm.conf^@SLURM_CPUS_ON_NODE=2^@SLURM_TASK_PID=1752643^@SLURM_NODEID=0^@SLURM_PROCID=0^@SLURM_LOCALID=0^@SLURM_GTIDS=0^@SLURM_JOB_UID=2574^@SLURM_JOB_USER=zeyer^@SLURM_WORKING_CLUSTER=cluster:mn-04:6817:9472:109^@SLURM_JOB_GID=2000^@SLURMD_NODENAME=cn-611^@`
</details>

The content of `/var/spool/slurm/slurmd/job9595321/slurm_script`:
```bash
#!/bin/sh
# This script was created by sbatch --wrap.

srun -o work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6/engine/%x.%A.%a /work/tools/users/zeyer/py-envs/py3.11-torch2.1/bin/python3.11 ./sis worker --engine long work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6 run
```

The env of 1752683 (first worker):
<details>

`SLURM_ARRAY_TASK_ID=1^@NCCL_DEBUG=INFO^@TMUX=/tmp/tmux-2574/default,1180403,2^@LANGUAGE=en_US.UTF-8^@LC_TIME=en_GB.UTF-8^@USER=zeyer^@SSH_CLIENT=137.226.36.158 33678 22^@SLURM_JOBID=9595321^@SLURM_JOB_USER=zeyer^@XDG_SESSION_TYPE=tty^@HOSTNAME=cn-611^@SLURM_JOB_QOS=normal^@LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu:/usr/local/cudnn-11.X-v8.4/lib64:/usr/local/cuda-11.7/lib64:/usr/local/cuda-11.7/extras/CUPTI/lib64:/usr/local/cudnn-10.1-v7.6/lib64:/usr/local/cudnn-9.1-v7.1/lib64:/usr/local/cudnn-8.0-v7.0/lib64:/usr/local/cudnn-8.0-v6.0/lib64:/usr/local/cudnn-8.0-v5.1/lib64:/usr/local/cuda-9.1/lib64:/usr/local/cuda-9.1/extras/CUPTI/lib64:/usr/local/cuda-10.1/lib64:/usr/local/cuda-10.1/extras/CUPTI/lib64:/usr/local/cuda-10.0/lib64:/usr/local/cuda-10.0/extras/CUPTI/lib64:/usr/lib/atlas-base:/usr/local/cuda-7.5/lib64^@SHLVL=2^@SLURM_JOB_NUM_NODES=1^@HOME=/u/zeyer^@MOTD_SHOWN=pam^@OLDPWD=/u/zeyer/setups/combined/2021-05-31^@SLURM_TASKS_PER_NODE=2^@SSH_TTY=/dev/pts/0^@TERM_PROGRAM_VERSION=3.3a^@ENVIRONMENT=BATCH^@SLURM_TOPOLOGY_ADDR_PATTERN=node^@SLURM_PRIO_PROCESS=0^@SLURM_ARRAY_TASK_MIN=1^@SLURM_WORKING_CLUSTER=cluster:mn-04:6817:9472:109^@LC_CTYPE=en_US.UTF-8^@LC_MONETARY=en_GB.UTF-8^@DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/2574/bus^@SLURM_JOB_NAME=i6_experiments.users.zeyer.datasets.utils.serialize.ReturnnDatasetToTextLinesJob.nVowScvUWQv6.run^@SLURM_ARRAY_TASK_MAX=1^@SLURM_JOB_CPUS_PER_NODE=2^@SLURM_JOB_GID=2000^@SLURM_JOB_ACCOUNT=hlt^@SLURM_CPUS_ON_NODE=2^@SLURM_PROCID=0^@INFOPATH=/u/zeyer/.linuxbrew/share/info:/u/zeyer/.linuxbrew/share/info:^@TMPDIR=/var/tmp^@SLURM_CONF=/var/spool/slurm/slurmd/conf-cache/slurm.conf^@LOGNAME=zeyer^@TEMP=/var/tmp^@SLURM_SUBMIT_HOST=indigo^@XDG_SESSION_CLASS=user^@TERM=tmux-256color^@XDG_SESSION_ID=10509^@SLURM_ARRAY_TASK_COUNT=1^@SLURM_NODELIST=cn-611^@SLURM_ARRAY_TASK_STEP=1^@SLURM_JOB_ID=9595321^@SLURM_NNODES=1^@HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1^@SLURM_JOB_NODELIST=cn-611^@SLURMD_NODENAME=cn-611^@PATH=/u/zeyer/.gem/ruby/1.9.1/bin:/u/zeyer/dotfiles/bin:/u/zeyer/dotfiles/system-tools/bin:/u/zeyer/tools/bin:/u/zeyer/.local/bin:/u/zeyer/.linuxbrew/bin:/u/zeyer/.gem/ruby/1.9.1/bin:/u/zeyer/dotfiles/bin:/u/zeyer/dotfiles/system-tools/bin:/u/zeyer/tools/bin:/u/zeyer/.local/bin:/u/zeyer/.linuxbrew/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolb
ox/scripts^@PAPERSIZE=a4^@SLURM_GTIDS=0,1^@LC_ADDRESS=en_GB.UTF-8^@XDG_RUNTIME_DIR=/run/user/2574^@SLURM_CPUS_PER_TASK=1^@LANG=en_US.UTF-8^@LC_TELEPHONE=en_GB.UTF-8^@TERM_PROGRAM=tmux^@LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=00:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.avif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:*~=00;90:*#=00;90:*.bak=00;90:*.old=00;90:*.orig=00;90:*.part=00;90:*.rej=00;90:*.swp=00;90:*.tmp=00;90:*.dpkg-dist=00;90:*.dpkg-old=00;90:*.ucf-dist=00;90:*.ucf-new=00;90:*.ucf-old=00;90:*.rpmnew=00;90:*.rpmorig=00;90:*.rpmsave=00;90:^@TMP=/var/tmp^@SLURM_CLUSTER_NAME=cluster^@SLURM_JOB_UID=2574^@SHELL=/bin/zsh^@LC_NAME=en_GB.UTF-8^@HOMEBREW_ARCH=core2^@SLURM_ARRAY_JOB_ID=9595321^@SLURM_LOCALID=0^@LC_MEASUREMENT=en_GB.UTF-8^@SLURM_JOB_PARTITION=cpu_modern^@LC_IDENTIFICATION=en_GB.UTF-8^@SLURM_TASK_PID=1752683^@PWD=/u/zeyer/setups/combined/2021-05-31^@LC_ALL=en_US.UTF-8^@SLURM_TOPOLOGY_ADDR=cn-611^@MY_LOCAL_MACHINE=indigo^@SSH_CONNECTION=137.226.36.158 33678 137.226.36.63 22^@SLURM_MEM_PER_NODE=20480^@LC_NUMERIC=en_GB.UTF-8^@HOMEBREW_FORCE_BREWED_CURL=1^@LC_PAPER=en_GB.UTF-8^@TMUX_PLUGIN_MANAGER_PATH=/u/zeyer/.tmux/plugins/^@MANPATH=/u/zeyer/.linuxbrew/share/man:/u/zeyer/.linuxbrew/share/man:^@TMUX_PANE=%13^@SLURM_SUBMIT_DIR=/u/zeyer/setups/combined/2021-05-31^@SLURM_NODE_ALIASES=(null)^@SLURM_NODEID=0^@SRUN_DEBUG=3^@SLURM_UMASK=0022^@SLURM_JOB_CPUS_PER_NODE_PACK_GROUP_0=2^@SLURM_NTASKS=2^@SLURM_NPROCS=2^@SLURM_DISTRIBUTION=block^@SLURM_STEP_ID=0^@SLURM_STEPID=0^@SLURM_SRUN_COMM_PORT=43041^@SLURM_STEP_NODELIST=cn-611^@SLURM_STEP_NUM_NODES=1^@SLURM_STEP_NUM_TASKS=2^@SLURM_STEP_TASKS_PER_NODE=2^@SLURM_STEP_LAUNCHER_PORT=43041^@SLURM_SRUN_COMM_HOST=10.6.13.11^@SLURM_LAUNCH_NODE_IPADDR=10.6.13.11^@`
</details>

The env of 1752684 (second worker):
<details>

`SLURM_ARRAY_TASK_ID=1^@NCCL_DEBUG=INFO^@TMUX=/tmp/tmux-2574/default,1180403,2^@LANGUAGE=en_US.UTF-8^@LC_TIME=en_GB.UTF-8^@USER=zeyer^@SSH_CLIENT=137.226.36.158 33678 22^@SLURM_JOBID=9595321^@SLURM_JOB_USER=zeyer^@XDG_SESSION_TYPE=tty^@HOSTNAME=cn-611^@SLURM_JOB_QOS=normal^@LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu:/usr/local/cudnn-11.X-v8.4/lib64:/usr/local/cuda-11.7/lib64:/usr/local/cuda-11.7/extras/CUPTI/lib64:/usr/local/cudnn-10.1-v7.6/lib64:/usr/local/cudnn-9.1-v7.1/lib64:/usr/local/cudnn-8.0-v7.0/lib64:/usr/local/cudnn-8.0-v6.0/lib64:/usr/local/cudnn-8.0-v5.1/lib64:/usr/local/cuda-9.1/lib64:/usr/local/cuda-9.1/extras/CUPTI/lib64:/usr/local/cuda-10.1/lib64:/usr/local/cuda-10.1/extras/CUPTI/lib64:/usr/local/cuda-10.0/lib64:/usr/local/cuda-10.0/extras/CUPTI/lib64:/usr/lib/atlas-base:/usr/local/cuda-7.5/lib64^@SHLVL=2^@SLURM_JOB_NUM_NODES=1^@HOME=/u/zeyer^@MOTD_SHOWN=pam^@OLDPWD=/u/zeyer/setups/combined/2021-05-31^@SLURM_TASKS_PER_NODE=2^@SSH_TTY=/dev/pts/0^@TERM_PROGRAM_VERSION=3.3a^@ENVIRONMENT=BATCH^@SLURM_TOPOLOGY_ADDR_PATTERN=node^@SLURM_PRIO_PROCESS=0^@SLURM_ARRAY_TASK_MIN=1^@SLURM_WORKING_CLUSTER=cluster:mn-04:6817:9472:109^@LC_CTYPE=en_US.UTF-8^@LC_MONETARY=en_GB.UTF-8^@DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/2574/bus^@SLURM_JOB_NAME=i6_experiments.users.zeyer.datasets.utils.serialize.ReturnnDatasetToTextLinesJob.nVowScvUWQv6.run^@SLURM_ARRAY_TASK_MAX=1^@SLURM_JOB_CPUS_PER_NODE=2^@SLURM_JOB_GID=2000^@SLURM_JOB_ACCOUNT=hlt^@SLURM_CPUS_ON_NODE=2^@SLURM_PROCID=1^@INFOPATH=/u/zeyer/.linuxbrew/share/info:/u/zeyer/.linuxbrew/share/info:^@TMPDIR=/var/tmp^@SLURM_CONF=/var/spool/slurm/slurmd/conf-cache/slurm.conf^@LOGNAME=zeyer^@TEMP=/var/tmp^@SLURM_SUBMIT_HOST=indigo^@XDG_SESSION_CLASS=user^@TERM=tmux-256color^@XDG_SESSION_ID=10509^@SLURM_ARRAY_TASK_COUNT=1^@SLURM_NODELIST=cn-611^@SLURM_ARRAY_TASK_STEP=1^@SLURM_JOB_ID=9595321^@SLURM_NNODES=1^@HOMEBREW_FORCE_BREWED_CA_CERTIFICATES=1^@SLURM_JOB_NODELIST=cn-611^@SLURMD_NODENAME=cn-611^@PATH=/u/zeyer/.gem/ruby/1.9.1/bin:/u/zeyer/dotfiles/bin:/u/zeyer/dotfiles/system-tools/bin:/u/zeyer/tools/bin:/u/zeyer/.local/bin:/u/zeyer/.linuxbrew/bin:/u/zeyer/.gem/ruby/1.9.1/bin:/u/zeyer/dotfiles/bin:/u/zeyer/dotfiles/system-tools/bin:/u/zeyer/tools/bin:/u/zeyer/.local/bin:/u/zeyer/.linuxbrew/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts:/u/zeyer/.local/share/JetBrains/Toolbox/scripts^@PAPERSIZE=a4^@SLURM_GTIDS=0,1^@LC_ADDRESS=en_GB.UTF-8^@XDG_RUNTIME_DIR=/run/user/2574^@SLURM_CPUS_PER_TASK=1^@LANG=en_US.UTF-8^@LC_TELEPHONE=en_GB.UTF-8^@TERM_PROGRAM=tmux^@LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=00:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.avif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:*~=00;90:*#=00;90:*.bak=00;90:*.old=00;90:*.orig=00;90:*.part=00;90:*.rej=00;90:*.swp=00;90:*.tmp=00;90:*.dpkg-dist=00;90:*.dpkg-old=00;90:*.ucf-dist=00;90:*.ucf-new=00;90:*.ucf-old=00;90:*.rpmnew=00;90:*.rpmorig=00;90:*.rpmsave=00;90:^@TMP=/var/tmp^@SLURM_CLUSTER_NAME=cluster^@SLURM_JOB_UID=2574^@SHELL=/bin/zsh^@LC_NAME=en_GB.UTF-8^@HOMEBREW_ARCH=core2^@SLURM_ARRAY_JOB_ID=9595321^@SLURM_LOCALID=1^@LC_MEASUREMENT=en_GB.UTF-8^@SLURM_JOB_PARTITION=cpu_modern^@LC_IDENTIFICATION=en_GB.UTF-8^@SLURM_TASK_PID=1752684^@PWD=/u/zeyer/setups/combined/2021-05-31^@LC_ALL=en_US.UTF-8^@SLURM_TOPOLOGY_ADDR=cn-611^@MY_LOCAL_MACHINE=indigo^@SSH_CONNECTION=137.226.36.158 33678 137.226.36.63 22^@SLURM_MEM_PER_NODE=20480^@LC_NUMERIC=en_GB.UTF-8^@HOMEBREW_FORCE_BREWED_CURL=1^@LC_PAPER=en_GB.UTF-8^@TMUX_PLUGIN_MANAGER_PATH=/u/zeyer/.tmux/plugins/^@MANPATH=/u/zeyer/.linuxbrew/share/man:/u/zeyer/.linuxbrew/share/man:^@TMUX_PANE=%13^@SLURM_SUBMIT_DIR=/u/zeyer/setups/combined/2021-05-31^@SLURM_NODE_ALIASES=(null)^@SLURM_NODEID=0^@SRUN_DEBUG=3^@SLURM_UMASK=0022^@SLURM_JOB_CPUS_PER_NODE_PACK_GROUP_0=2^@SLURM_NTASKS=2^@SLURM_NPROCS=2^@SLURM_DISTRIBUTION=block^@SLURM_STEP_ID=0^@SLURM_STEPID=0^@SLURM_SRUN_COMM_PORT=43041^@SLURM_STEP_NODELIST=cn-611^@SLURM_STEP_NUM_NODES=1^@SLURM_STEP_NUM_TASKS=2^@SLURM_STEP_TASKS_PER_NODE=2^@SLURM_STEP_LAUNCHER_PORT=43041^@SLURM_SRUN_COMM_HOST=10.6.13.11^@SLURM_LAUNCH_NODE_IPADDR=10.6.13.11^@`
</details>

**Comment by albertz** [comment]
Some info about the job:
```
$ scontrol show job 9595321  
JobId=9595321 ArrayJobId=9595321 ArrayTaskId=1 JobName=i6_experiments.users.zeyer.datasets.utils.serialize.ReturnnDatasetToTextLinesJob.nVowScvUWQv6.run
   UserId=zeyer(2574) GroupId=assi(2000) MCS_label=N/A
   Priority=606 Nice=0 Account=hlt QOS=normal
   JobState=RUNNING Reason=None Dependency=(null)
   Requeue=1 Restarts=0 BatchFlag=1 Reboot=0 ExitCode=0:0
   RunTime=02:59:20 TimeLimit=2-02:00:00 TimeMin=N/A
   SubmitTime=2024-12-03T10:11:40 EligibleTime=2024-12-03T10:11:41
   AccrueTime=2024-12-03T10:11:41
   StartTime=2024-12-03T10:11:43 EndTime=2024-12-05T12:11:43 Deadline=N/A
   PreemptEligibleTime=2024-12-03T10:11:43 PreemptTime=None
   SuspendTime=None SecsPreSuspend=0 LastSchedEval=2024-12-03T10:11:43 Scheduler=Backfill
   Partition=cpu_modern AllocNode:Sid=indigo:1181078
   ReqNodeList=(null) ExcNodeList=cn-[222,231,240,251,253,257]
   NodeList=cn-611
   BatchHost=cn-611
   NumNodes=1 NumCPUs=2 NumTasks=1 CPUs/Task=1 ReqB:S:C:T=0:0:*:*
   TRES=cpu=2,mem=20G,node=1
   Socks/Node=* NtasksPerN:B:S:C=0:0:*:* CoreSpec=*
   MinCPUsNode=1 MinMemoryNode=20G MinTmpDiskNode=0
   Features=(null) DelayBoot=00:00:00
   OverSubscribe=OK Contiguous=0 Licenses=(null) Network=(null)
   Command=(null)
   WorkDir=/u/zeyer/setups/combined/2021-05-31
   StdErr=/u/zeyer/setups/combined/2021-05-31/work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6/engine/i6_experiments.users.zeyer.datasets.utils.serialize.ReturnnDatasetToTextLinesJob.nVowScvUWQv6.run.9595321.1.batch
   StdIn=/dev/null
   StdOut=/u/zeyer/setups/combined/2021-05-31/work/i6_experiments/users/zeyer/datasets/utils/serialize/ReturnnDatasetToTextLinesJob.nVowScvUWQv6/engine/i6_experiments.users.zeyer.datasets.utils.serialize.ReturnnDatasetToTextLinesJob.nVowScvUWQv6.run.9595321.1.batch
   Power=
```

**Comment by albertz** [comment]
The `rqmt` that I used for this job are `{"cpu": 1, "mem": 20, "time": 50, "gpu": 0}`. Specifically, the `submit_log.run`:
`([1], {'cpu': 1, 'mem': 20.0, 'time': 50.0, 'sbatch_args': ['-x', 'cn-257,cn-251,cn-253,cn-231,cn-240,cn-222'], 'gpu': 0, 'engine_info': [([], ['9595321'])], 'engine_name': 'slurm'})`

**Comment by albertz** [comment]
It seems it anyway requested 2 CPUs? Why?

But that shouldn't be so much the problem. But then, it seems `SLURM_NTASKS` and `SLURM_NPROCS` are also set to 2, to the number of CPUs? So those are the wrong env vars? We should use `SLURM_NNODES` instead?

**Comment by albertz** [comment]
But again, why are there two workers running here?

**Comment by JackTemaki** [comment]
> It seems it anyway requested 2 CPUs? Why?
> 
> But that shouldn't be so much the problem. But then, it seems `SLURM_NTASKS` and `SLURM_NPROCS` are also set to 2, to the number of CPUs? So those are the wrong env vars? We should use `SLURM_NNODES` instead?

This one I can answer. Slurm does not allow to split physical cores between two jobs. As our nodes have hyperthreading enabled, each physical core corresponds to 2 virtual/hyper cores. Thus, Slurm will always internally increase `cpu` to the next even number.

**Comment by NeoLegends** [comment]
> Why are there two procs (1752684 and 1752683)?

It is very strange to me that you get `SLURM_NTASKS=2` and `SLURM_NPROCS=2` in your job when even SLURM says the job has only one task (`NumNodes=1, NumTasks=1`).

> We should use SLURM_NNODES instead?

I think we should first find out why you get two tasks spawned. You seem to be getting at least coherent `SLURM_PROCID`s (0 and 1) in your worker processes. 

Does this happen consistently to your jobs? Is there anything else (scripts, env vars from .bashrc/.zshrc etc.) besides the obvious messing with your jobs?

**Comment by albertz** [comment]
The `SLURM_NTASKS` only appear in the workers but not in the initial slurm_script. There however, I see `SLURM_TASKS_PER_NODE=2`. But where does this come from?

> Does this happen consistently to your jobs?

Yes, it seems so.

>  Is there anything else (scripts, env vars from .bashrc/.zshrc etc.) besides the obvious messing with your jobs?

I haven't changed anything in the last year or so on this.

**Comment by albertz** [comment]
> But then the question is, where does `SLURM_CPUS_PER_TASK=1` come from?

That comes from us:
```python
        out.append("--cpus-per-task=%s" % rqmt.get("cpu", 1))
```

**Comment by albertz** [comment]
@JackTemaki:

> Thus, Slurm will always internally increase `cpu` to the next even number.

How exactly do you do that? Can you also increase `SLURM_CPUS_PER_TASK` in the same way? I think that is the problem here.

**Comment by JackTemaki** [comment]
> @JackTemaki:
> 
> > Thus, Slurm will always internally increase `cpu` to the next even number.
> 
> How exactly do you do that? Can you also increase `SLURM_CPUS_PER_TASK` in the same way? I think that is the problem here.

I never did anything, this is a Slurm default (https://slurm.schedmd.com/slurm.conf.html#OPT_SelectTypeParameters):

> CR_Core_Memory
> Cores and memory are consumable resources. On nodes with hyper-threads, each thread is counted as a CPU to satisfy a job's resource requirement, but multiple jobs are not allocated threads on the same core. The count of CPUs allocated to a job may be rounded up to account for every CPU on an allocated core. Setting a value for DefMemPerCPU is strongly recommended. 

We might get different behaviour when switching to e.g. `CR_CPU_Memory`, but this is something I will not experiment with this year any more. The implications are not clear, and I do not want to conduct multiple controller restarts again and possibly making our cluster unstable just for testing this, So you have to figure out how to fix it on the Sisyphus side.

> Btw, wouldn't everyone at i6 have this problem?

Probably nobody (including me) updated Sisyphus since #212 then.

**Comment by albertz** [comment]
> > Btw, wouldn't everyone at i6 have this problem?
> 
> Probably nobody (including me) updated Sisyphus since #212 then.

I'm not just talking about the issue with the logfiles. I'm talking about the issue that it runs two tasks here. I think that would be the case for all jobs where you request an uneven number of CPUs. Independent of the Sisyphus version.

**Comment by JackTemaki** [comment]
> You don't consider this a serious issue that it runs two tasks?

I did not say this, I am saying this is not happening for me, or for most other people it seems.

**Comment by albertz** [comment]
> > You don't consider this a serious issue that it runs two tasks?
> 
> I did not say this, I am saying this is not happening for me, or for most other people it seems.

Are you sure? You might not have noticed this. Can you run any job with rqmt={cpu:1} and check how many tasks this runs? Usually you will not get any error then. The output file might also be just fine depending on how the job runs (and just random luck).

**Comment by albertz** [comment]
So, I did this simple test:
`slurm-test-script.sh` content:
```
#!/bin/sh
#SBATCH --time=1
srun hostname |sort
```
Run `sbatch --cpus-per-task=1 slurm-test-script.sh`.
Log file:
```
slurm-9608616.out                                                                                                         
--------------------- Slurm Task Prolog ------------------------                                                                              
Job ID: 9608616                                                                                                                               
Job name: slurm-test-script.sh                                                                                                                
Host: cn-603
Date: Tue Dec  3 04:02:13 PM CET 2024
User: zeyer             
Slurm account: hlt      
Slurm partition: cpu_modern
Work dir:         
------------------         
Node usage:                
JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON) 
9608616 cpu_moder slurm-te    zeyer  R       0:01      1 cn-603 
------------------                                                     
Show launch script with:                                               
sacct -B -j 
------------------
--------------------- Slurm Task Prolog ------------------------
--------------------- Slurm Task Epilog ------------------------
Job ID: 9608616                                                        
Time: Tue Dec  3 04:02:14 PM CET 2024
Elapsed Time: 00:00:02                                                 
Billing per second for TRES: cpu=1,mem=190000M,node=1
Show resource usage with e.g.:                                         
sacct -j 9608616 -o Elapsed,TotalCPU,UserCPU,SystemCPU,MaxRSS,ReqTRES%60,MaxDiskRead,MaxDiskWrite
--------------------- Slurm Task Epilog ------------------------                                                                              
--------------------- Slurm Task Epilog ------------------------
Job ID: 9608616                                                        
Time: Tue Dec  3 04:02:14 PM CET 2024    
Elapsed Time: 00:00:02
Billing per second for TRES: cpu=1,mem=190000M,node=1
Show resource usage with e.g.:
sacct -j 9608616 -o Elapsed,TotalCPU,UserCPU,SystemCPU,MaxRSS,ReqTRES%60,MaxDiskRead,MaxDiskWrite
--------------------- Slurm Task Epilog ------------------------
------------------
------------------
------------------
------------------
------------------
------------------
9608616 cpu_moder slurm-te    zeyer  R       0:02      1 cn-603
9608616 cpu_moder slurm-te    zeyer  R       0:02      1 cn-603
cn-603
cn-603
Date: Tue Dec  3 04:02:14 PM CET 2024
Date: Tue Dec  3 04:02:14 PM CET 2024
Host: cn-603
Host: cn-603
Job ID: 9608616
Job ID: 9608616
JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON) 
JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON) 
Job name: slurm-test-script.sh
Job name: slurm-test-script.sh
Node usage:
Node usage:
sacct -B -j 
sacct -B -j 
Show launch script with:
Show launch script with:
Slurm account: hlt
Slurm account: hlt                 
Slurm partition: cpu_modern
Slurm partition: cpu_modern
--------------------- Slurm Task Prolog ------------------------
--------------------- Slurm Task Prolog ------------------------
--------------------- Slurm Task Prolog ------------------------
--------------------- Slurm Task Prolog ------------------------
User: zeyer
User: zeyer
Work dir: 
Work dir: 
--------------------- Slurm Task Epilog ------------------------
Job ID: 9608616
Time: Tue Dec  3 04:02:14 PM CET 2024
Elapsed Time: 00:00:03
Billing per second for TRES: cpu=1,mem=190000M,node=1
Show resource usage with e.g.:
sacct -j 9608616 -o Elapsed,TotalCPU,UserCPU,SystemCPU,MaxRSS,ReqTRES%60,MaxDiskRead,MaxDiskWrite
--------------------- Slurm Task Epilog ------------------------
```

So, it seems to affect everyone.

**Comment by JackTemaki** [comment]
I am not sure why, I just submitted a KenLM job with cpu=1, and there is only one worker and one process on cn-810

**Comment by albertz** [comment]
Can you try with my example?

**Comment by albertz** [comment]
I have a potential fix in PR #229.

**Comment by JackTemaki** [comment]
> Can you try with my example?

Your example does something different than pre- #212 Sisyphus for submission. I am not doubting at all the result of your script, I am doubting that this was an issue with older Sisyphus versions. I just wanted to check the

> So, it seems to affect everyone.

Statement. 

So I would see this as regression introduced with #212

**Comment by albertz** [comment]
Ah I thought my script was doing what Sisyphus was always doing (because I just copy&pasted the first example from the Slurm `sbatch` man page without really thinking about it).

So, this additional `srun` call seems to make the difference? But shouldn't it be always correct to have this? Also, it seems actually reasonable to me that we get `SLURM_TASKS_PER_NODE=2` when we have `SLURM_CPUS_PER_TASK=1` and `NumCPUs=2`.

In any case, PR #230 should fix this.

**Comment by albertz** [comment]
Ok, it needs some more investigation. With higher (uneven) cpu rqmt, the behavior is again different (only a single task?).

I extended my example script as:
```bash
#!/bin/sh
#SBATCH --time=1

echo "Base script"
env

echo "Now via SLURM srun"
srun env
```
And then ran `sbatch --cpus-per-task=3 slurm-test-script.sh`.

In the log, grepped for `SLURM`, I get:
<details>

```
SLURM_JOBID=9610349
SLURM_JOB_USER=zeyer
SLURM_JOB_QOS=normal
SLURM_JOB_NUM_NODES=1
SLURM_TASKS_PER_NODE=1
SLURM_TOPOLOGY_ADDR_PATTERN=node
SLURM_PRIO_PROCESS=0
SLURM_WORKING_CLUSTER=cluster:mn-04:6817:9472:109
SLURM_JOB_NAME=slurm-test-script.sh
SLURM_JOB_CPUS_PER_NODE=4
SLURM_JOB_GID=2000
SLURM_JOB_ACCOUNT=hlt
SLURM_CPUS_ON_NODE=4
SLURM_PROCID=0
SLURM_CONF=/var/spool/slurm/slurmd/conf-cache/slurm.conf
SLURM_SUBMIT_HOST=indigo
SLURM_NODELIST=cn-603
SLURM_JOB_ID=9610349
SLURM_NNODES=1
SLURM_JOB_NODELIST=cn-603
SLURMD_NODENAME=cn-603
SLURM_GTIDS=0
SLURM_CPUS_PER_TASK=3
SLURM_CLUSTER_NAME=cluster
SLURM_JOB_UID=2574
SLURM_LOCALID=0
SLURM_JOB_PARTITION=cpu_modern
SLURM_TASK_PID=1199374
SLURM_TOPOLOGY_ADDR=cn-603
SLURM_MEM_PER_NODE=190000
SLURM_SUBMIT_DIR=/u/zeyer
SLURM_NODE_ALIASES=(null)
SLURM_NODEID=0
Now via SLURM srun
SLURM_JOBID=9610349
SLURM_JOB_USER=zeyer
SLURM_JOB_QOS=normal
SLURM_JOB_NUM_NODES=1
SLURM_TASKS_PER_NODE=1
SLURM_TOPOLOGY_ADDR_PATTERN=node
SLURM_PRIO_PROCESS=0
SLURM_WORKING_CLUSTER=cluster:mn-04:6817:9472:109
SLURM_JOB_NAME=slurm-test-script.sh
SLURM_JOB_CPUS_PER_NODE=4
SLURM_JOB_GID=2000
SLURM_JOB_ACCOUNT=hlt
SLURM_CPUS_ON_NODE=4
SLURM_PROCID=0
SLURM_CONF=/var/spool/slurm/slurmd/conf-cache/slurm.conf
SLURM_SUBMIT_HOST=indigo
SLURM_NODELIST=cn-603
SLURM_JOB_ID=9610349
SLURM_NNODES=1
SLURM_JOB_NODELIST=cn-603
SLURMD_NODENAME=cn-603
SLURM_GTIDS=0
SLURM_CPUS_PER_TASK=3
SLURM_CLUSTER_NAME=cluster
SLURM_JOB_UID=2574
SLURM_LOCALID=0
SLURM_JOB_PARTITION=cpu_modern
SLURM_TASK_PID=1199394
SLURM_TOPOLOGY_ADDR=cn-603
SLURM_MEM_PER_NODE=190000
SLURM_SUBMIT_DIR=/u/zeyer
SLURM_NODE_ALIASES=(null)
SLURM_NODEID=0
SLURM_UMASK=0022
SLURM_JOB_CPUS_PER_NODE_PACK_GROUP_0=4
SLURM_NTASKS=1
SLURM_NPROCS=1
SLURM_DISTRIBUTION=cyclic
SLURM_STEP_ID=0
SLURM_STEPID=0
SLURM_SRUN_COMM_PORT=35705
SLURM_STEP_NODELIST=cn-603
SLURM_STEP_NUM_NODES=1
SLURM_STEP_NUM_TASKS=1
SLURM_STEP_TASKS_PER_NODE=1
SLURM_STEP_LAUNCHER_PORT=35705
SLURM_SRUN_COMM_HOST=10.6.13.3
SLURM_LAUNCH_NODE_IPADDR=10.6.13.3
```
</details>
So it only runs once.

**Comment by JackTemaki** [comment]
From the documentation:

> --ntasks-per-node=<ntasks>
>     Request that ntasks be invoked on each node. If used with the --ntasks option, the --ntasks option will take precedence and the --ntasks-per-node will be treated as a maximum count of tasks per node. Meant to be used with the --nodes option. This is related to --cpus-per-task=ncpus, but does not require knowledge of the actual number of cpus on each node. In some cases, it is more convenient to be able to request that no more than a specific number of tasks be invoked on each node. Examples of this include submitting a hybrid MPI/OpenMP app where only one MPI "task/rank" should be assigned to each node while allowing the OpenMP portion to utilize all of the parallelism present in the node, or submitting a single setup/cleanup/monitoring job to each node of a pre-existing allocation as one step in a larger job script. 

Especially the

> "This is related to --cpus-per-task=ncpus, but does not require knowledge of the actual number of cpus on each node"

Maybe this already solves our problem?

**Comment by albertz** [comment]
So you suggest that we additionally add `--ntasks-per-node=1`? With `sbatch --cpus-per-task=1 --ntasks-per-node=1 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=1`.

But I still don't fully understand the logic of `SLURM_TASKS_PER_NODE`. With `sbatch --cpus-per-task=1 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=2`, but with `sbatch --cpus-per-task=3 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=1`.

**Comment by JackTemaki** [comment]
> So you suggest that we additionally add `--ntasks-per-node=1`? With `sbatch --cpus-per-task=1 --ntasks-per-node=1 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=1`.
> 
> But I still don't fully understand the logic of `SLURM_TASKS_PER_NODE`. With `sbatch --cpus-per-task=1 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=2`, but with `sbatch --cpus-per-task=3 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=1`.

If I understand correctly, `cpus-per-task` can be dropped then

**Comment by JackTemaki** [comment]
> So you suggest that we additionally add `--ntasks-per-node=1`? With `sbatch --cpus-per-task=1 --ntasks-per-node=1 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=1`.
> 
> But I still don't fully understand the logic of `SLURM_TASKS_PER_NODE`. With `sbatch --cpus-per-task=1 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=2`, but with `sbatch --cpus-per-task=3 slurm-test-script.sh`, I get `SLURM_TASKS_PER_NODE=1`.

Yes exactly

**Comment by RaphaelLilt** [comment]
I'm also having the issue that jobs are running twice. @albertz Could we reopen this issue?

**Comment by albertz** [comment]
> I'm also having the issue that jobs are running twice. @albertz Could we reopen this issue?

You mean still with the latest master, i.e. with #231 merged? Because since we have #231, I thought this is fixed?

**Comment by RaphaelLilt** [comment]
Yes, I'm using the latest commit on `main`: [a0319f3](https://github.com/rwth-i6/sisyphus/commit/a0319f31160646ae1601acd5fc58bf9883402936)
I'll try https://github.com/rwth-i6/sisyphus/pull/230

**Comment by RaphaelLilt** [comment]
I'm also seeing some weird behavior in the logs, where a job is in `running` and then back in `runnable` state:
```
256904:[2024-12-10 00:48:41,749] INFO: runnable: Job<xxxx.V7CRuW6LvUfX>
257333:[2024-12-10 00:48:42,364] INFO: Submit to queue: xxxx.V7CRuW6LvUfX split [1]
257335:[2024-12-10 00:48:42,372] INFO: Submitted with job_id: ['4408036'] xxxx.split
257413:[2024-12-10 00:49:21,008] INFO: queue: Job<xxxx.V7CRuW6LvUfX>
257916:[2024-12-10 00:49:32,011] INFO: running: Job<xxxxV7CRuW6LvUfX>
258345:[2024-12-10 00:49:40,700] INFO: running: Job<xxxx.V7CRuW6LvUfX>
258819:[2024-12-10 00:50:07,524] INFO: runnable: Job<xxxx.V7CRuW6LvUfX>
259207:[2024-12-10 00:50:07,886] INFO: Submit to queue: xxxx.V7CRuW6LvUfX split [1]
259213:[2024-12-10 00:50:07,894] INFO: Submitted with job_id: ['4408430'] xxxx.V7CRuW6LvUfX.split
259268:[2024-12-10 00:50:14,070] INFO: queue: Job<xxxx.V7CRuW6LvUfX>
259825:[2024-12-10 00:50:19,685] INFO: runnable: Job<xxx.V7CRuW6LvUfX>
```

**Comment by michelwi** [comment]
Could you maybe also run @albertz example script above and share the env(s) you get?
Any information about your slurm install would probably also be helpful if you can share it.

**Comment by albertz** [comment]
@RaphaelLilt I think this is a different issue. Can you open a separate issue with that log and whatever other information you can share about your Slurm env? I think it's different because in my case, I never saw anything like that on the Sisyphus manager log. The Sisyphus manager log was totally correct. Only when checking on the node, while the job was running, I noticed that two procs were actually running, not just one. This was also only for CPU jobs where rqmt had cpu=1. I also noticed the wrong/broken symlinks of the `log` files in the job dir.

**Comment by michelwi** [comment]
Yes, actually now that Albert mentions it. The manager log to me looks like Sisyphus worker has problems starting on the compute node. 
If the worker does not start correctly, no logfiles are linked and the worker process terminates without the manager thread noticing that it has ever run, resulting in state runnable again.

In this case check the `xxxx.V7CRuW6LvUfX/engine/` folder for any slurm logfiles which should contain an error message. Include this if any in case you open a new issue. Thx.
